src/skin/tooltipupdater.cpp

```diff
@@ -0,0 +1,170 @@
+#include "widget/wslidercomposed.h"
+#include "widget/wpushbutton.h"
+#include "tooltipupdater.h"
+
+TooltipShortcutUpdater::TooltipShortcutUpdater() { }
+
+TooltipShortcutUpdater::~TooltipShortcutUpdater() {
+    // TODO(Tomasito) Delete WidgetTooltipWatchers
+}
+
+void TooltipShortcutUpdater::addWatcher(QList<ConfigKey> configKeys, WBaseWidget* pWidget) {
+    WSliderComposed* sliderComposed = toSliderComposed(pWidget);
+    WPushButton* pushButton = toPushButton(pWidget);
+
+    if (sliderComposed) {
+        m_pWatchers.append(new SliderTooltipWatcher(
+                &m_pKbdPreset,
+                 configKeys,
+                 sliderComposed
+        ));
+    } else if (pushButton) {
+        m_pWatchers.append(new PushButtonTooltipWatcher(
+                &m_pKbdPreset,
+                configKeys,
+                pushButton
+        ));
+    } else {
+        m_pWatchers.append(new WidgetTooltipWatcher(
+                &m_pKbdPreset,
+                configKeys,
+                pWidget
+        ));
+    }
+}
+
+
+void TooltipShortcutUpdater::updateShortcuts(ControllerPresetPointer pPreset) {
+    QSharedPointer<KeyboardControllerPreset> pKbdPreset = pPreset.dynamicCast<KeyboardControllerPreset>();
```

Comment:
Yes sir :) I got rid of all this casts from `QSharedPointer<ControllerPreset>` to `QSharedPointer<KeyboardControllerPreset>` by making KeyboardController emit a new signal `keyboardControllerPresetLoaded(QSharedPointer<KeyboardControllerPreset>` (typedefed as KeyboardControllerPresetPointer), so there is no need anymore for casting.

Eh.. now that I am thinking about it. I could have probably just overloaded the presetLoaded signal receiving a `KeyboardControllerPreset` instead of `ControllerPreset`.. at least .., is signal overloading possible?